File: backend/src/main/java/com/soccersignup/backend/service/TeamSheetService.java
package com.soccersignup.backend.service;

import com.soccersignup.backend.dto.TeamSheetRequest;
import com.soccersignup.backend.dto.TeamSheetResponse;
import com.soccersignup.backend.model.Game;
import com.soccersignup.backend.model.Player;

public interface TeamSheetService {

    TeamSheetResponse getTeamSheet(Long gameId);

    TeamSheetResponse autoSplit(Long gameId, Long expectedVersion);

    TeamSheetResponse saveTeamSheet(Long gameId, TeamSheetRequest request);

    TeamSheetResponse publishTeamSheet(Long gameId, Long expectedVersion);

    void handlePublishedSheetDeparture(
            Game game,
            Player leavingPlayer,
            Player promotedPlayer);
}

File: frontend/soccer-frontend/src/app/services/team-sheet.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { TeamSheet, TeamSheetRequest } from '../models/team-sheet.model';

@Injectable({ providedIn: 'root' })
export class TeamSheetService {

  constructor(private http: HttpClient) {}

  private url(gameId: number): string {
    return `/api/games/${gameId}/teamsheet`;
  }

  // Get the team sheet for a game
  // Returns 404 if not published and caller is not an organiser
  getTeamSheet(gameId: number): Observable<TeamSheet> {
    return this.http.get<TeamSheet>(this.url(gameId));
  }

  // Ask the backend to randomly split confirmed players into two teams
  autoSplit(gameId: number): Observable<TeamSheet> {
    return this.http.post<TeamSheet>(`${this.url(gameId)}/auto-split`, {});
  }

  // Save the full team sheet (organiser's drag-and-drop result)
  saveTeamSheet(gameId: number, request: TeamSheetRequest): Observable<TeamSheet> {
    return this.http.put<TeamSheet>(this.url(gameId), request);
  }

  // Publish the team sheet — notifies all confirmed players
  publishTeamSheet(gameId: number): Observable<TeamSheet> {
    return this.http.post<TeamSheet>(`${this.url(gameId)}/publish`, {});
  }
}
